src/mscorlib/src/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -49,38 +73,91 @@ public abstract class AssemblyLoadContext
         internal static extern void InternalStartProfile(string profile, IntPtr ptrNativeAssemblyLoadContext);
 #endif // FEATURE_MULTICOREJIT
 
-        protected AssemblyLoadContext()
+        static AssemblyLoadContext()
         {
-            // Initialize the ALC representing non-TPA LoadContext
-            InitializeLoadContext(false);
+            // We register the cleanup of all AssemblyLoadContext that have not been finalized in the AppContext.Unloading
+            AppContext.Unloading += OnAppContextUnloading;
         }
 
-        internal AssemblyLoadContext(bool fRepresentsTPALoadContext)
+#if FEATURE_COLLECTIBLE_ALC
+        protected AssemblyLoadContext() : this(false, false)
         {
-            // Initialize the ALC representing TPA LoadContext
-            InitializeLoadContext(fRepresentsTPALoadContext);
         }
-        
-        void InitializeLoadContext(bool fRepresentsTPALoadContext)
+
+        protected AssemblyLoadContext(bool isCollectible) : this(false, isCollectible)
```

Comment:
@gkhanna79, since this constructor will be in the System.Runtime.Loader contract, should it also exist when `FEATURE_COLLECTIBLE_ALC` is not defined? I'm assuming the contract can't be conditional on the feature, so then it wouldn't compile in CoreFX when the feature is turned off. Same for `Unload` and `IsCollectible`. Could probably throw `PlatformNotSupportedException` when the feature is disabled.